### PR TITLE
feat: Add Git Hook for Ruff

### LINKED ISSUE #47: [Feature Request] Add Git hook for Ruff
### Overview

Add Git hook for Ruff.

### Details

Add pre-commit hook for Ruff for `ruff format` and `ruff check`.

### Why would this feature be useful?

To improve the development experience by automatically formatting and linting code if possible.

### Additional Notes

_No response_

### PR DESCRIPTION
**Pull request by SouthernPolaris**
### Description
Adds auto activation of ruff when a commit is made. Automatically performs a ruff --fix (which fixes both the "check" and "format" options at once)

### Changes Made
Added .pre-commit-config.yaml file

### Related Issues
Fixes #47: "[Feature Request] Add Git hook for Ruff"

### CODE REVIEW

File: src/data_fetcher.py
Comment (phoenixpereira) [inline src/data_fetcher.py:1]: You should cherry-pick the commits to exclude the logging functionality commits as that's another PR

File: .pre-commit-config.yaml
Comment (phoenixpereira) [inline .pre-commit-config.yaml:1]: You need to add pre-commit as a dev dependency using uv.  Also, you'll need to update the README to include running `pre-commit install` to set the hook up unless there's a better way that only requires you to run`uv sync` to set it up.

**Review by phoenixpereira — CHANGES_REQUESTED** [review]

**Comment by SouthernPolaris** [comment]
Force pushed after pulling from upstream main to ensure it is not dependent on other pull request

File: .pre-commit-config.yaml
Comment (SouthernPolaris) [reply .pre-commit-config.yaml:1]: For now, I updated README, but I will look into uv sync to set it up automatically

File: README.md [hunk 1]
```diff
@@ -44,6 +44,22 @@ uv run python3 src/scraper.py
...
+To perform a commit from local, pre-commit must first be installed:
+```
+pip3 install pre-commit
+```
+OR if performing the action in bash:
+```
+python3 -m pip install pre-commit
+```
+
+Then, as recommended by [Official Pre-Commit Documentation](https://pre-commit.com/):
```
Comment (phoenixpereira) [inline README.md:57]: Reword to better highlight what this command is doing.
```suggestion
Install the git hook scripts:
```

File: README.md [hunk 2]
```diff
@@ -44,6 +44,22 @@ uv run python3 src/scraper.py
...
+```
+pre-commit install
+```
+This should be run every time you clone a project
```
Comment (phoenixpereira) [inline README.md:61]: Not really relevant.
```suggestion
```

File: README.md [hunk 3]
```diff
@@ -44,6 +44,22 @@ uv run python3 src/scraper.py
...
 
+This repository is configured with a pre-commit linting and formatting tool
+To perform a commit from local, pre-commit must first be installed:
+```
+pip3 install pre-commit
+```
+OR if performing the action in bash:
+```
+python3 -m pip install pre-commit
+```
```
Comment (phoenixpereira) [inline README.md:55]: We want to have pre-commit installed at a project level, so you can just do `uv add pre-commit --dev` and pre-commit should be installed when you do `uv sync`. You still need to do `pre-commit install` though to install the hook.

File: README.md [hunk 4]
```diff
@@ -44,6 +44,22 @@ uv run python3 src/scraper.py
 
 We welcome contributions to enhance Courses API! If you find any issues, have suggestions, or want to request a feature, please follow our [Contributing Guidelines](https://github.com/compsci-adl/.github/blob/main/CONTRIBUTING.md).
 
+This repository is configured with a pre-commit linting and formatting tool
```
Comment (phoenixpereira) [inline README.md:47]: Reword to better highlight what the pre-commit hook does.
```suggestion
This repository is configured with a pre-commit hook for automated linting and formatting.
```

**Review by phoenixpereira — CHANGES_REQUESTED** [review]

**Review by phoenixpereira — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+repos:
+- repo: https://github.com/astral-sh/ruff-pre-commit
+  # Ruff version
+  rev: v0.7.3
+  hooks:
+    # Run the linter
+    - id: ruff
+      args: [ --fix ]
+    # Run the formatter
+    - id: ruff-format
```
Comment (rayokamoto) [inline .pre-commit-config.yaml:10]:
```suggestion
    - id: ruff-format

```

**Review by rayokamoto — CHANGES_REQUESTED** [review]

**Review by rayokamoto — APPROVED** [review]

**Comment by rayokamoto** [comment]
> Implements Issue 47: "[Feature Request] Add Git hook for Ruff"

BTW, you can link issues to a pull request which will automatically resolve them once the PR is merged. [See more here](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue).

**Pull request merged**